src/Interpreters/Streaming/Aggregator.cpp

```diff
@@ -1310,52 +1307,205 @@ Block Aggregator::convertOneBucketToBlockFinal(AggregatedDataVariants & variants
+        std::set<Int64> buckets_set;
+        for (auto & data_variants : non_empty_data)
+        {
+            auto tmp_buckets = getDataVariant<Method>(*data_variants).data.buckets();
+            buckets_set.insert(tmp_buckets.begin(), tmp_buckets.end());
+        }
+        buckets.assign(buckets_set.begin(), buckets_set.end());
+    }
+
+    std::unique_ptr<ThreadPool> thread_pool;
+    auto num_threads = std::min(max_threads, buckets.size());
+    if (num_threads > 1)
+        thread_pool = std::make_unique<ThreadPool>(num_threads);
+
+    /// proton FIXME : separate final vs non-final converting. For non-final converting, we don't need
+    /// each arena for each thread.
+    for (size_t i = first.aggregates_pools.size(); i < num_threads; ++i)
+        first.aggregates_pools.push_back(std::make_shared<Arena>());
+
+    std::atomic<size_t> next_bucket_idx_to_merge = 0;
+
+    auto converter = [&](size_t thread_id, ThreadGroupStatusPtr thread_group) {
+        SCOPE_EXIT_SAFE(if (thread_group) CurrentThread::detachQueryIfNotDetached(););
+        if (thread_group)
+            CurrentThread::attachToIfDetached(thread_group);
+
+        BlocksList blocks;
+        while (true)
+        {
+            UInt32 bucket_idx = next_bucket_idx_to_merge.fetch_add(1);
+            if (bucket_idx >= buckets.size())
+                break;
+
+            auto bucket = buckets[bucket_idx];
+
+            /// Merge one bucket into first one
+            Arena * arena = first.aggregates_pools.at(thread_id).get();
+            mergeBucketImpl<Method>(non_empty_data, final, clear_states, bucket, arena);
+            auto & method = getDataVariant<Method>(first);
+            if (method.data.impls[bucket].empty())
+                continue;
+
+            /// Convert one bucket of first one
+            blocks.emplace_back(convertOneBucketToBlockImpl(first, method, arena, final, clear_states, bucket));
+        }
+
+        if (clear_states)
+        {
+            for (size_t i = 1; i < non_empty_data.size(); ++i)
+                clearDataVariants(*non_empty_data[i]);
+        }
+
+        return blocks;
+    };
+
+    /// packaged_task is used to ensure that exceptions are automatically thrown into the main stream.
+    std::vector<std::packaged_task<BlocksList()>> tasks(num_threads);
+    try
+    {
+        for (size_t thread_id = 0; thread_id < num_threads; ++thread_id)
+        {
+            tasks[thread_id] = std::packaged_task<BlocksList()>(
+                [group = CurrentThread::getGroup(), thread_id, &converter] { return converter(thread_id, group); });
+
+            if (thread_pool)
+                thread_pool->scheduleOrThrowOnError([thread_id, &tasks] { tasks[thread_id](); });
+            else
+                tasks[thread_id]();
+        }
+    }
+    catch (...)
+    {
+        /// If this is not done, then in case of an exception, tasks will be destroyed before the threads are completed, and it will be bad.
```

Comment:
The other exception-safe trick is creating a `std::shared_ptr<BlocksList> results` and each thread fill its own results in the corresponding index / slot, in the driving thread, we just collect the results. Something like below

ThreadPool ctor / creation can be deferred as late as possible and separate concurrent merge and single thread merge sounds make the logic much clearer and compact.

```
if (num_threads <= 1)
    return converter();

ThreadPool> thread_pool(num_threads);

auto results = std::make_shared<std::vector<BlockList>();
results->resize(num_threads);

for (size_t thread_id = 0; thread_id < num_threads; ++thread_id)
    thread_pool->scheduleOrThrowOnError([thread_id, group = CurrentThread::getGroup(), results] { 
        (*results)[thread_id] = converter(thread_id, group);
    });

thread_pool.wait();

BlocksList blocks;
for (auto & result : *results)
    blocks.splice(blocks.end(), result);

return blocks;
```